### PR TITLE
[MM-25553] favicon added for unread messages without mention

### PR DESCRIPTION
**Pull request by bhargav50**
#### Summary

A new favicon is introduced when there are unread messages, but there is no mention

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/15400

#### Changes
- favicon component changed and enum is introduced to keep track of what to show instead of boolean which was used previously
- favicon component unit tests are modified to check exact enum value instead of checking true / false

### Unread Messages, but no mention

image

### CODE REVIEW

**Review by hmhealey — APPROVED** [review]
Awesome, looks great to me! I appreciate that it's making it clear that something has to change for the favicon to be updated in the `componentDidUpdate` and that you removed some repeated code in `updateFavicon` as well.

**Review by devinbinnie — APPROVED** [review]
LGTM! Thanks @bhargav50!

**Comment by jgilliam17** [comment]
Thank you @bhargav50 
This is great, new favicon color looks awesome!
Question
I verified that * has been removed when unreads are present, should we remove the * for mentions?
Screen Shot 2020-09-18 at 3 46 57 PM

**Comment by esethna** [comment]
@jgilliam17 yes we should remove it for mentions as well. Can you also help test on various browsers/operating systems?

**Comment by bhargav50** [comment]
@esethna shall I remove * for mentions in this PR?

**Comment by esethna** [comment]
@bhargav50 yes please do.

**Comment by hmhealey** [comment]
Hold on, we still need the * for browsers that don't support changing the Favicon, so that has to be re-added for those

**Comment by jgilliam17** [comment]
> Hold on, we still need the * for browsers that don't support changing the Favicon, so that has to be re-added for those

@hmhealey Is this only impacting Safari on macOS?

**Comment by jgilliam17** [comment]
Tested on Safari 14.0 (previous versions are not displaying favicons)
- It appears that favicon is not changing when unreads/mentions are present 
Screen Shot 2020-09-24 at 10 24 19 AM

**Comment by hmhealey** [comment]
It was needed for Safari and Edge previously. I don't know if Edge's support for it changed with it switching to Chromium, but the code explicitly only enables it on Firefox and Chrome currently.

And for reference, Edge supported showing the favicon previously, but it had some weird caching behaviour, so the favicon wasn't updated reliably when the user received or read mentions.

**Comment by jgilliam17** [comment]
Tested on New Edge (chromium based, macOS and Windows) and the favicons display as expected. 
Screen Shot 2020-09-24 at 2 22 09 PM
Screenshot 2020-09-24 142828
Screenshot 2020-09-24 142444 (1)

**Comment by bhargav50** [comment]
Shall I add * again then? :sweat_smile:

**Comment by jgilliam17** [comment]
> Shall I add * again then? 😅

Based on testing, I would recommend we have the `*` available on Safari since the favicon is not updating
cc: @esethna

**Comment by esethna** [comment]
@bhargav50 apologies, I should have been more clear. On browsers that support the custom favicons we should remove the *, but for browsers that don't we should keep the *

**Comment by bhargav50** [comment]
@hmhealey * is added on unread messages in safari browser, one unit test is also added to assert this behavior in safari

**Review by esethna — APPROVED** [review]

**Review by jgilliam17 — APPROVED** [review]
Thank you @bhargav50 
Tested, looks good to merge. 
- Verified unread and mention custom favicons on Chrome, Edge and Firefox on macOS 
- Verified unread and mention custom favicons on Chrome, Edge and Firefox on Windows 
- Verified unread and mention custom favicons on Firefox on Ubuntu.
- Verified on Safari(macOS) main favicon color/style has been updated. We are continuing to use asterisk for both unreads and mentions, due to favicon not updating on Safari.
Screen Shot 2020-09-29 at 5 24 17 PM

**Review by hmhealey — CHANGES_REQUESTED** [review]
Also, would you mind reverting the changes to package-lock.json to fix the conflict with master?

File: components/favicon_title_handler/favicon_title_handler.tsx [hunk 1]
```diff
@@ -83,28 +91,56 @@ class FaviconTitleHandler extends React.PureComponent<Props> {
                     currentChannelName = currentTeammate.display_name;
                 }
             }
-
             const mentionTitle = unreads.mentionCount > 0 ? '(' + unreads.mentionCount + ') ' : '';
-            const unreadTitle = unreads.messageCount > 0 ? '* ' : '';
+            const unreadTitle = UserAgent.isSafari() && unreads.messageCount > 0 ? '* ' : '';
```
[the remark this answers is not in the document]
Comment (bhargav50) [reply components/favicon_title_handler/favicon_title_handler.tsx]: I have moved this check to a getter and checked for firefox and chrome to disable updating * in the title

**Review by hmhealey — APPROVED** [review]
Thanks for changing that

**Pull request merged**